File: leetcode/matrix/leetcode-305-number-of-islands-ii.java
// LeetCode 305. Number of Islands II
// https://leetcode.com/problems/number-of-islands-ii/
//

class Solution {
    public List<Integer> numIslands2(int m, int n, int[][] positions) {
        int[][] grid = new int[m][n];
        int[] parents = new int[m * n];
        int[] ranks = new int[m * n];
        for (int i = 0; i < m * n; i++) parents[i] = i;
        int[][] delta = new int[][] { {1, 0}, {-1, 0}, {0, 1}, {0, -1} };
        int count = 0;
        List<Integer> list = new ArrayList();

        for (int[] pos: positions) {
            int val = pos[0] * n + pos[1];
            if (grid[pos[0]][pos[1]] != 1) {
                grid[pos[0]][pos[1]] = 1;
                count++;
            }
            for (int d = 0; d < delta.length; d++) {
                int r = pos[0] + delta[d][0];
                int c = pos[1] + delta[d][1];
                if (inBound(m, n, r, c) && grid[r][c] == 1) {
                    if (unionSet(val, r * n + c, parents, ranks)) count--;
                }
            }
            list.add(count);
        }
        return list;
    }

    private int findParent(int val, int[] parents, int[] ranks) {
        while (val != parents[val]) val = parents[val];
        return val;
    }

    private boolean unionSet(int val1, int val2, int[] parents, int[] ranks) {
        int parent1 = findParent(val1, parents, ranks);
        int parent2 = findParent(val2, parents, ranks);
        if (parent1 == parent2) return false;

        if (ranks[val1] > ranks[val2]) {
            parents[parent2] = parent1;
        } else if (ranks[val1] < ranks[val2]) {
            parents[parent1] = parent2;
        } else {
            parents[parent1] = parent2;
            ranks[val2]++;
        }
        return true;
    }

    private boolean inBound(int m, int n, int r, int c) {
        return r >= 0 && r < m && c >= 0 && c < n;
    }
}

File: leetcode/matrix/leetcode-305-number-of-islands-ii.rb
# LeetCode 305. Number of Islands II
# https://leetcode.com/problems/number-of-islands-ii/
#

class Solution
    def solve(m, n, positions)
        grid = Array.new(m) { Array.new(n, 0) }
        parents = Array.new(m * n)
        ranks = Array.new(m * n, 0)

        list = []
        islands = 0
        deltas = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        positions.each do |r, c|
            val = r * n + c
            if grid[r][c] == 0
                grid[r][c] = 1
                islands += 1
                parents[val] = val
            end

            deltas.each do |dr, dc|
                if in_bound?(grid, r + dr, c + dc) && grid[r + dr][c + dc] == 1
                    islands -= 1 if union(val, (r + dr) * n + c + dc, parents, ranks)
                end
            end

            list.push(islands)
        end
        list
    end

    private

    def in_bound?(grid, r, c)
        r >= 0 && r < grid.size && c >= 0 && c < grid[0].size
    end

    def find_parent(val, parents)
        while parents[val] != val
            # with path compression
            parents[val] = parents[parents[val]]
            val = parents[val]
        end
        val
    end

    def union(val1, val2, parents, ranks)
        parent1 = find_parent(val1, parents)
        parent2 = find_parent(val2, parents)

        return false if parent1 == parent2

        if ranks[parent1] > ranks[parent2]
            parents[parent2] = parent1
        elsif ranks[parent1] < ranks[parent2]
            parents[parent1] = parent2
        else
            parents[parent1] = parent2
            ranks[parent2] += 1
        end
        true
    end
end

# @param {Integer} m
# @param {Integer} n
# @param {Integer[][]} positions
# @return {Integer[]}
def num_islands2(m, n, positions)
    Solution.new.solve(m, n, positions)
end
